src/box/lua/config/utils/odict.lua

```diff
@@ -0,0 +1,209 @@
+-- Ordered dictionary.
+--
+-- The collection tracks the order in which the items are added
+-- and provides odict.pairs() function to get them in this order.
+--
+-- The ordered dictionary is a usual Lua table with a specific
+-- metatable. All the table operations are applicable.
+--
+-- It is similar to Python's collections.OrderedDict.
+--
+-- Example:
+--
+--  | local od = odict.new()
+--  |
+--  | od.a = 1
+--  | od.b = 2
+--  | od.c = 3
+--  |
+--  | print('od.a', od.a) -- 1
+--  | print('od.b', od.b) -- 2
+--  | print('od.c', od.c) -- 3
+--  |
+--  | for k, v in odict.pairs(od) do
+--  |     print(k, v)
+--  | end
+--  | -- print: a, 1
+--  | -- print: b, 2
+--  | -- print: c, 3
+--
+-- If an element is deleted and added again, it is added to the
+-- end.
+--
+--  | local od = odict.new()
+--  |
+--  | od.a = 1
+--  | od.b = 2
+--  | od.c = 3
+--  |
+--  | od.b = nil
+--  | od.b = 4
+--  |
+--  | for k, v in odict.pairs(od) do
+--  |     print(k, v)
+--  | end
+--  | -- print: a, 1
+--  | -- print: c, 3
+--  | -- print: b, 4
+--
+-- If an element is changed (without prior deletion), it remains
+-- on the same position.
+
+local REINDEX_THRESHOLD = 1000
+
+local registry = setmetatable({}, {__mode = 'k'})
+
+-- Generator function for odict.pairs().
+--
+-- Returns (key, value).
+local function gen(od, prev_key)
+    local ctx = registry[od]
+
+    -- The previous key is nil only on the first call of the
+    -- generator function.
+    local id = prev_key == nil and 0 or ctx.key2id[prev_key]
+    -- NB: This assert doesn't catch all the kinds of changes
+    -- during an iteration. It rather verifies a precondition
+    -- for the following cycle.
+    assert(id ~= nil, 'ordered dictionary is changed during iteration')
+
+    while id <= ctx.max_id do
+        id = id + 1
+        local key = ctx.id2key[id]
+        -- id2key may have holes for IDs <= max_id in place of
+        -- deleted items.
+        --
+        -- id2key may contain a stalled entry, because __newindex
+        -- is not called on assignment of an existing field,
+        -- including assignment to nil.
+        if key ~= nil and od[key] ~= nil then
+            return key, od[key]
+        end
+    end
+end
+
+-- Returns a Lua iterator triplet (gen, param, state).
+--
+-- Usage:
+--
+--  | for k, v in odict.pairs(od) do
+--  |     <...>
+--  | end
+--
+-- The iterator stability guarantees are the same as for usual
+-- pairs().
+--
+-- Quote from the Lua 5.1 reference manual:
+--
+-- > The behavior of `next` is undefined if, during the traversal,
+-- > you assign any value to a non-existent field in the table.
+-- > You may however modify existing fields. In particular, you
+-- > may clear existing fields.
+local function _pairs(od)
+    return gen, od, nil
+end
+
+-- Delete the key from the id<->key mappings.
+local function trash(ctx, key)
```

Comment:
I would prefer a different function name here, because "to trash something" has a different meaning. Maybe: `remove_key(ctx, key)` or `drop_key(ctx, key)`

Reply:
I've started from `set` and `drop` functions, but than found that I need functions to control tracking of the keys separately from assignments into the key-value table.

The idea of the `trash` and `track` names are that they have no association with the key-value table and dedicated only to the internal tracking structures. IOW, they have no `drop`/`delete`/`remove` and `set`/`add`/`append` words. Also, `trash` reminds the `TRASH` macro from `src/trivia/util.h`.

I like these names, but I would hear thoughts from other reviewers before a decision to keep or change them.

Reply:
Renamed.

```diff
diff --git a/src/box/lua/config/utils/odict.lua b/src/box/lua/config/utils/odict.lua
index dce3e9ad52..ffefb00ec0 100644
--- a/src/box/lua/config/utils/odict.lua
+++ b/src/box/lua/config/utils/odict.lua
@@ -108,7 +108,7 @@ local function _pairs(od)
 end
 
 -- Delete the key from the id<->key mappings.
-local function trash(ctx, key)
+local function release(ctx, key)
     local id = ctx.key2id[key]
     if id ~= nil then
         ctx.key2id[key] = nil
@@ -149,7 +149,7 @@ local function reindex(od, ctx)
         local key = ctx.id2key[id]
         -- Drop the given key-id pair from the key<->id mappings.
         if key ~= nil then
-            trash(ctx, key)
+            release(ctx, key)
         end
         -- Add the new key-id pair into the key<->id mappings.
         if key ~= nil and od[key] ~= nil then
@@ -168,7 +168,7 @@ local mt = {
 
         -- If a field with this key had existed, we should update
         -- the key<->id mappings to let gen() know the new order.
-        trash(ctx, k)
+        release(ctx, k)
 
         -- If a non-existing field is assigned to nil, we have
         -- nothing to do.
```